momaland/learning/iql/iql.py

```diff
@@ -0,0 +1,273 @@
+
+    Args:
+        weights: the weights for the objectives
+        rewards: the rewards for the objectives
+
+    Returns:
+        float: the utility of the action
+    """
+    return np.dot(weights, rewards)
+
+
+class QAgent:
+    """Q-learning agent."""
+
+    def __init__(self, agent_id, n_states, n_actions):
+        """Initialize the Q-agent."""
+        self.agent_id = agent_id
+        self.n_states = n_states
+        self.n_actions = n_actions
+        self.q_values = np.zeros((n_states, n_actions), dtype=np.float32)
+
+    def act(self, state, epsilon):
+        """Epsilon-greedy action selection.
+
+        Choose the action with the highest Q-value with probability 1-epsilon otherwise choose a random action.
+        """
+        if np.random.random() < epsilon:
+            return np.random.randint(0, self.n_actions)
+        else:
+            return np.argmax(self.q_values[state])
+
+    def update(self, state, new_state, action, alpha, gamma, reward, done):
+        """Update the Q-values of the agent based on the chosen action and the new state."""
+        # Retrieve the current Q-value
+        q_value = self.q_values[state][action]
+        # Compute the next max Q-value
+        next_max_q_value = 0.0 if done else np.max(self.q_values[new_state])
+        self.q_values[state][action] = q_value + alpha * (reward + gamma * next_max_q_value - q_value)
+
+
+def train(args, weights, env_args):
+    """IQL scalarizing the vector reward using weights and weighted sum."""
+    # Environment Initialization
+    env = TabularMOBeachDomainWrapper(**env_args)
+    obs, infos = env.reset()
+
+    agents = {
+        QAgent(agent_id, env.observation_space(env.agents[0]).n, env.action_spaces[agent_id].n)
+        for agent_id in env.possible_agents
+    }
+
+    # Algorithm specific parameters
+    epsilon = args.epsilon
+    epsilon_decay = args.epsilon_decay
+    epsilon_min = args.epsilon_min
+    alpha = args.alpha
+    alpha_decay = args.alpha_decay
+    alpha_min = args.alpha_min
+    gamma = args.gamma
+
+    episode_returns = []
+    # keep track of the best reward encountered
+    best_reward = np.array([-np.inf, -np.inf])
+    best_reward_scal = -np.inf
+
+    for current_iter in range(args.num_iterations):
+        # Get the actions from the agents
+        if args.random:
+            actions = {q_agent.agent_id: random.choice(range(env.action_spaces[q_agent.agent_id].n)) for q_agent in agents}
+        else:
+            actions = {q_agent.agent_id: q_agent.act(obs[q_agent.agent_id], epsilon) for q_agent in agents}
+
+        # Update the exploration rate
+        if epsilon > epsilon_min:
+            epsilon *= epsilon_decay
+        else:
+            epsilon = epsilon_min
+
+        new_obs, rew, terminateds, truncations, infos = env.step(actions)
+
+        # Check if all agents are terminated (in a stateless setting, this is always the case)
```

Comment:
Can you add a "TODO handle truncation and termination difference"

Reply:
Done